### PR TITLE
feat: add paged loading for contacts

### PR DESCRIPTION
**Pull request by beyond009**
## Description
https://mask.atlassian.net/browse/MF-4989
https://mask.atlassian.net/browse/MF-4990
https://mask.atlassian.net/browse/MF-5001

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/popups/hook/useFriends.ts [hunk 1]
```diff
@@ -43,90 +44,94 @@ export const PlatformSort: Record<NextIDPlatform, number> = {
...
+                        pageOffset: 0,
+                    },
+                    3000,
+                )
+                if (values.length === 0) return EMPTY_LIST
+                cachedValues.current = values
+            }
+            const friends: Friend[] = []
+            const startIndex = pageParam ? 10 * Number(pageParam) : 0
+            for (let i = startIndex; i < cachedValues.current.length; i += 1) {
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/hook/useFriends.ts]:
```ts
const friends = cachedValues.map(() => {
})
```

Comment (beyond009) [reply packages/mask/src/extension/popups/hook/useFriends.ts]: Need to break the loop when there are 10 valid results

File: packages/mask/src/extension/popups/hook/useLoadMore.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { useState, useEffect, type RefObject } from 'react'
+import { throttle } from 'lodash-es'
+export function useLoadMore(ref: RefObject<HTMLElement>) {
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/hook/useLoadMore.ts]: CC @UncleBill  Do we have any pre-existing solution?

Comment (UncleBill) [reply packages/mask/src/extension/popups/hook/useLoadMore.ts]: We do.
@beyond009 You can take a look at this:

https://github.com/DimensionDev/Maskbook/blob/dd35db783d5b6446107242b8c1040e51a0c5cab9/packages/mask/src/extension/popups/pages/Wallet/components/ActivityList/index.tsx#L78-L80

File: packages/mask/src/extension/popups/hook/useFriends.ts [hunk 3]
```diff
@@ -43,90 +44,94 @@ export const PlatformSort: Record<NextIDPlatform, number> = {
...
+                    currentPersona?.identifier,
+                    {
+                        network: 'all',
+                        pageOffset: 0,
+                    },
+                    3000,
+                )
+                if (values.length === 0) return EMPTY_LIST
+                cachedValues.current = values
+            }
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hook/useFriends.ts]: You can split into multiple queries, and control the order by `enabled`

https://tanstack.com/query/v4/docs/react/guides/dependent-queries


```ts
    // Get records first
    const { data: records = EMPTY_LIST, isLoading } = useQuery(['relation-records'], async () => {
        return Services.Identity.queryRelationPaged(
            currentPersona?.identifier,
            {
                network: 'all',
                pageOffset: 0,
            },
            3000,
        )
    })
    const allProfiles = useMemo(() => records.map((x) => x.profile), [records])
    useInfiniteQuery({
        enabled: !isLoading,  // When `records` get fetched, this query will begin.
        queryKey: ['friends', currentPersona?.identifier.rawPublicKey, allProfiles],
        queryFn: async ({ pageParam }) => {
            const offset = pageParam || 0
            const profile = allProfiles[offset]
            if (isProfileIdentifier(profile)) {
                const friend = first(await Services.Identity.queryProfilesInformation([profile]))
                return {
                    friend,
                    nextOffset: offset + 1,
                }
            } else {
                return {
                    friend: { persona: profile },
                    nextOffset: offset + 1,
                }
            }
            // You an query multiple profile each page.
            // const chunk = 10
            // const profiles = allProfiles.slice(offset, offset + chunk)
            // const promises = profiles.map(x => { GET_PROFILES_BY_CHUNK })
            // return {
            //   friends: await promises
            //   nextOffset: nextOffset + chunk
            // }
        },
        getNextPageParam: (lastPage) => {
            if (lastPage?.nextOffset >= allProfiles.length) return
            return lastPage?.nextOffset
        },
    })

```

Comment (UncleBill) [reply packages/mask/src/extension/popups/hook/useFriends.ts]: I also implemented an infinite query cross multiple chains.

https://github.com/DimensionDev/Maskbook/blob/dd35db783d5b6446107242b8c1040e51a0c5cab9/packages/web3-hooks/base/src/useNonFungibleAssets.ts#L24-L47

File: packages/mask/src/extension/popups/pages/Friends/Contacts/index.tsx [hunk 4]
```diff
@@ -56,6 +56,7 @@ export const Contacts = memo<ContactsProps>(function Contacts({ friends, listRef
                     />
                 )
             })}
+            <ElementAnchor callback={() => fetchNextPage} height={10} />
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/pages/Friends/Contacts/index.tsx]:
```suggestion
            <ElementAnchor callback={() => fetchNextPage()} height={10} />
```

File: packages/mask/src/extension/popups/hook/useFriends.ts [hunk 5]
```diff
@@ -43,90 +42,95 @@ export const PlatformSort: Record<NextIDPlatform, number> = {
...
-        })
-        return uniqBy(profiles, ({ id }) => id)
-    }, [currentPersona])
+                filtered.sort((a, b) => PlatformSort[a.platform] - PlatformSort[b.platform])
+                return {
+                    profiles: filtered,
+                    ...friends[index],
+                    id: friends[index].persona.publicKeyAsHex,
+                }
+            })
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hook/useFriends.ts]: You should move data reconstructing out outside this query.

Comment (guanbinrui) [reply packages/mask/src/extension/popups/hook/useFriends.ts]: CC @beyond009

File: packages/mask/src/extension/popups/hook/useFriends.ts [hunk 6]
```diff
@@ -43,90 +42,95 @@ export const PlatformSort: Record<NextIDPlatform, number> = {
...
-            filtered.sort((a, b) => PlatformSort[a.platform] - PlatformSort[b.platform])
-            return {
-                profiles: filtered,
-                ...friends[index],
-                id: friends[index].persona.publicKeyAsHex,
-            }
-        })
-        return uniqBy(profiles, ({ id }) => id)
-    }, [currentPersona])
+                filtered.sort((a, b) => PlatformSort[a.platform] - PlatformSort[b.platform])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/hook/useFriends.ts]: Make it chainable.

```suggestion
               .sort((a, b) => PlatformSort[a.platform] - PlatformSort[b.platform])
```

File: packages/mask/src/extension/popups/pages/Friends/Home/index.tsx [hunk 7]
```diff
@@ -21,16 +22,19 @@ const FriendsHome = memo(function FriendsHome() {
         false,
     )
     const { value: searchedList = EMPTY_LIST } = useFriendsFromSearch(searchResult, value, _value)
+    const fetchMore = useCallback(() => {
+        hasNextPage ? fetchNextPage() : null
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Friends/Home/index.tsx:26]: if hasNextPage
    fetchNextPage

**Pull request merged**